edb/pgsql/delta.py

```diff
@@ -2258,12 +2258,142 @@ def adjust_pointer_storage(self, pointer, schema, orig_schema, context):
 
             self.pgops.add(opg)
 
+    def _alter_pointer_optionality(
+        self,
+        schema: s_schema.Schema,
+        orig_schema: s_schema.Schema,
+        context: sd.CommandContext,
+    ) -> None:
+        new_required = self.scls.get_required(schema)
+
+        ptr = self.scls
+        ptr_stor_info = types.get_pointer_storage_info(ptr, schema=schema)
+        ptr_table = ptr_stor_info.table_type == 'link'
+        is_lprop = ptr.is_link_property(schema)
+        is_multi = ptr_table and not is_lprop
+        is_required = ptr.get_required(schema)
+
+        if is_multi:
+            if isinstance(self, sd.AlterObjectFragment):
+                source_op = self.get_parent_op(context)
+            else:
+                source_op = self
+        else:
+            source_ctx = self.get_referrer_context_or_die(context)
+            source_op = source_ctx.op
+
+        # Ignore optionality changes resulting from the creation of
+        # an overloaded pointer as there is no data yet.
+        if isinstance(source_op, sd.CreateObject):
+            return
+
+        if self.fill_expr is not None:
+            _, fill_sql_expr, orig_rel_alias, _ = (
+                self._compile_conversion_expr(
+                    pointer=ptr,
+                    conv_expr=self.fill_expr,
+                    schema=schema,
+                    orig_schema=orig_schema,
+                    context=context,
+                    orig_rel_is_always_source=True,
+                )
+            )
+
+            if is_lprop:
+                obj_id_ref = f'{qi(orig_rel_alias)}.source'
+            else:
+                obj_id_ref = f'{qi(orig_rel_alias)}.id'
+
+            if is_required and not is_multi:
+                fill_sql_expr = textwrap.dedent(f'''\
+                    edgedb.raise_on_null(
+                        ({fill_sql_expr}),
+                        'not_null_violation',
+                        msg => 'missing value for required property',
+                        detail => '{{"object_id": "' || {obj_id_ref} || '"}}',
+                        "column" => {ql(str(ptr.id))}
+                    )
+                ''')
+
+            tab = q(*ptr_stor_info.table_name)
+            target_col = ptr_stor_info.column_name
+
+            if not is_multi:
+                update_qry = textwrap.dedent(f'''\
+                    UPDATE {tab} AS {qi(orig_rel_alias)}
+                    SET {qi(target_col)} = ({fill_sql_expr})
+                    WHERE {qi(target_col)} IS NULL
+                ''')
+                self.pgops.add(dbops.Query(update_qry))
+            else:
+                source = ptr.get_source(orig_schema)
+                src_tab = q(*common.get_backend_name(
+                    orig_schema,
+                    source,
+                    catenate=False,
+                ))
+
+                update_qry = textwrap.dedent(f'''\
```

Comment:
For this, do we create a new link table and then delete the old one?

Reply:
No, this simply inserts the result of evaluation of the `USING` clause into a link table for every object that has an empty link set.  I'll clarify in a comment.